File: src/main/java/com/ocbc/auctionservice/controllers/UserController.java
package com.ocbc.auctionservice.controllers;

import com.ocbc.auctionservice.authentications.UserContext;
import com.ocbc.auctionservice.configurations.RequestSynchronizationManager;
import com.ocbc.auctionservice.controllers.requests.UserRequest;
import com.ocbc.auctionservice.controllers.responses.UserResponse;
import com.ocbc.auctionservice.entities.User;
import com.ocbc.auctionservice.services.UserService;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@RestController
@RequestMapping("/api/v1/users")
public class UserController {

    @Autowired
    private UserService userService;

    @Autowired
    private RequestSynchronizationManager requestSynchronizationManager;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @GetMapping
    public ResponseEntity<List<UserResponse>> getUsers() {
        log.info("Request for all users for " + requestSynchronizationManager.loggedInUsername());
        return ResponseEntity.ok(userService.getUsers().stream().map(
                this::userToUserResponse).collect(Collectors.toList()));
    }

    @GetMapping("/{id}")
    public ResponseEntity<UserResponse> getUser(@PathVariable int id) {
        return ResponseEntity.ok(userToUserResponse(
                userService.getUser(id)));
    }

    @PostMapping
    public ResponseEntity<UserResponse> createUser(@Valid @RequestBody UserRequest userRequest) {
        return ResponseEntity.ok(userToUserResponse(
                userService.createUser(userRequestToUser(userRequest))));
    }

    @PutMapping("/{id}")
    public ResponseEntity<UserResponse> updateUser(@Valid @RequestBody UserRequest userRequest, @PathVariable int id) {
        return ResponseEntity.ok(userToUserResponse(
                userService.updateUser(userRequestToUser(userRequest), id)));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<String> deleteUser(@PathVariable int id) {
        userService.deleteUser(id);
        return ResponseEntity.ok("success");
    }

    private User userRequestToUser(UserRequest userRequest) {
        return User.builder()
                .name(userRequest.getName())
                .age(userRequest.getAge())
                .address(userRequest.getAddress())
                .password(passwordEncoder.encode(userRequest.getPassword()))
                .roles(userRequest.getRoles())
                .permissions(userRequest.getPermissions())
                .username(userRequest.getName().toLowerCase())
                .build();
    }

    private UserResponse userToUserResponse(User user) {
        return UserResponse.builder()
                .address(user.getAddress())
                .id(user.getId())
                .name(user.getName())
                .age(user.getAge())
                .build();
    }
}

File: src/main/kotlin/com/auction/service/controllers/UserController.kt
package com.auction.service.controllers

import com.auction.service.controllers.requests.UserRequest
import com.auction.service.controllers.responses.UserResponse
import com.auction.service.entities.User
import com.auction.service.services.UserService
import org.springframework.http.ResponseEntity
import org.springframework.http.ResponseEntity.ok
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*

@Controller
@RequestMapping("/api/v1/users")
class UserController(
        private val userService: UserService
) {

    @GetMapping // READ
    fun getUsers(): ResponseEntity<List<UserResponse>> = ok(userService.getUsers().map { user -> toUserResponse(user) })

    @GetMapping("/{id}") // READ
    fun getUser(@PathVariable id: Int): ResponseEntity<UserResponse> =
            ok(toUserResponse(userService.getUser(id)))

    @GetMapping("/{username}/username") // READ
    fun getUser(@PathVariable username: String): ResponseEntity<UserResponse> =
            ok(toUserResponse(userService.getUser(username)))

    @PostMapping // CREATE
    fun createUser(@RequestBody userRequest: UserRequest): ResponseEntity<UserResponse> =
            ok(toUserResponse(userService.createUser(toUser(userRequest))))

    @PutMapping("/{id}") // UPDATE
    fun updateUser(@RequestBody userRequest: UserRequest, @PathVariable id: Int): ResponseEntity<UserResponse> =
            ok(toUserResponse(userService.updateUser(toUser(userRequest), id)))

    @DeleteMapping("/{id}") // DELETE
    fun deleteUser(@PathVariable id: Int): ResponseEntity<String> {
        userService.deleteUser(id)
        return ok("ok")
    }

    private fun toUser(userRequest: UserRequest): User{
        return User().apply {
            name = userRequest.name
            username = userRequest.username
            address = userRequest.address
            password = userRequest.password
            age = userRequest.age
            roles = userRequest.roles
        }
    }

    private fun toUserResponse(user: User): UserResponse{
        return UserResponse(
                user.id!!,
                user.name,
                user.username,
                user.address,
                user.age!!,
                user.roles,
                user.permission)
    }
}